addons/udes_stock_picking_batch/models/stock_picking_batch.py

```diff
@@ -1,5 +1,581 @@
+        picking is not provided.
+        """
+        Picking = self.env["stock.picking"]
+        PickingBatch = self.env["stock.picking.batch"]
+
+        self.ensure_one()
+
+        reserve_pallet_per_picking = self.picking_type_ids.u_reserve_pallet_per_picking
+        if reserve_pallet_per_picking:
+            if not picking:
+                raise ValidationError(
+                    "A picking must be specified if pallets are reserved per picking."
+                )
+        if reserve_pallet_per_picking and picking and not self.picking_ids & picking:
+            raise ValidationError("Picking %s is not in batch %s." % (picking.name, self.name))
+
+        if reserve_pallet_per_picking:
+            conflicting_picking = Picking.search(
+                [
+                    ("u_reserved_pallet", "=", pallet_name),
+                    ("state", "not in", ["draft", "cancel", "done"]),
+                ]
+            )
+            if conflicting_picking:
+                raise ValidationError(
+                    _("This pallet is already being used for picking %s.")
+                    % conflicting_picking[0].name
+                )
+        else:
+            conflicting_batch = PickingBatch.search(
+                [
+                    ("id", "!=", self.id),
+                    ("state", "=", "in_progress"),
+                    ("u_last_reserved_pallet_name", "=", pallet_name),
+                ]
+            )
+            if conflicting_batch:
+                raise ValidationError(
+                    _("This pallet is already being used for batch %s.") % conflicting_batch[0].name
+                )
+
+        if reserve_pallet_per_picking:
+            # The front end will always send only one picking.
+            picking.write({"u_reserved_pallet": pallet_name})
+        else:
+            self.write({"u_last_reserved_pallet_name": pallet_name})
+
+    def _get_task_grouping_criteria(self):
+        """
+        Return a function for sorting by picking, package(maybe), product, and
+        location. The package is not included if the picking type allows for
+        the swapping of packages (`u_allow_swapping_packages`) and picks by
+        product (`u_user_scans`)
+        """
+        batch_pt = self.mapped("picking_ids.picking_type_id")
+        batch_pt.ensure_one()
+
+        parts = [lambda ml: (ml.picking_id.id,)]
+
+        if not (batch_pt.u_allow_swapping_packages and batch_pt.u_user_scans == "product"):
+            parts.append(lambda ml: (ml.package_id.id,))
+
+        parts.append(lambda ml: (ml.location_id.id, ml.product_id.id))
+
+        return lambda ml: tuple(chain(*[part(ml) for part in parts]))
+
+    def get_available_move_lines(self):
+        """ Get all the move lines from available pickings
+        """
+        self.ensure_one()
+        available_pickings = self.picking_ids.filtered(lambda p: p.state == "assigned")
+
+        return available_pickings.move_line_ids
```

Comment:
This would probably automatically sort the move lines by the ordering of the pickings (creating a JOIN behind the scenes) which in some cases it is not needed, usually we have some code to sort the move lines after calling this function; which can be inefficient with big batches.
Can we add a parameter `sort = False` and use `.search(XXX, order="id")` when it is false, please?

Reply:
Resolving, as this is done and we have another thread on how to name the method. :)